File: backend/helpers/MockUsersGen/mock_users_gen.go
package MockUsersGen

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"eunity.com/backend-main/helpers/DBManager"
	"eunity.com/backend-main/models"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

func get_random_user(amount string) ([]models.User, error) {
	// URL of the API endpoint
	url := "https://randomuser.me/api/?results=" + fmt.Sprint(amount)

	// Make a GET request to the endpoint
	resp, err := http.Get(url)
	if err != nil {
		log.Fatalf("Error making GET request: %v", err)
	}
	defer resp.Body.Close()

	// Check the response status code
	if resp.StatusCode != http.StatusOK {
		log.Fatalf("Error: Unexpected status code %d", resp.StatusCode)
	}

	// Decode the JSON response
	var data struct {
		Results []map[string]interface{} `json:"results"`
	}
	err = json.NewDecoder(resp.Body).Decode(&data)
	if err != nil {
		log.Fatalf("Error decoding JSON: %v", err)
	}

	var users []models.User
	for _, userData := range data.Results {
		objectId := primitive.NewObjectID()

		// Parse date of birth
		dateString := userData["dob"].(map[string]interface{})["date"].(string)
		parsedDate, err := time.Parse(time.RFC3339, dateString)
		if err != nil {
			fmt.Println("Error parsing date:", err)
			return nil, err
		}
		dob := models.DateOfBirth{
			Day:   parsedDate.Day(),
			Month: int(parsedDate.Month()),
			Year:  parsedDate.Year(),
		}

		user := models.User{
			ID:             &objectId,
			Email:          userData["email"].(string),
			Verified_email: true,
			PhoneNumber:    userData["phone"].(string),
			FirstName:      userData["name"].(map[string]interface{})["first"].(string),
			LastName:       userData["name"].(map[string]interface{})["last"].(string),
			Gender:         userData["gender"].(string),
			Location:       userData["location"].(map[string]interface{})["country"].(string),
			MatchPreferences: models.MatchPreferences{
				Genders:           []string{"Men", "Women"},
				RelationshipTypes: []string{"Long Term Relationships"},
				MinimumAge:        18,
				MaximumAge:        39,
				MaximumDistance:   40,
			},
			DateOfBirth: &dob,
			Height: &models.Height{
				Feet:   5,
				Inches: 8,
			},
			Providers: map[string]models.Provider{
				"google": {
					Name:           userData["name"].(map[string]interface{})["first"].(string) + " " + userData["name"].(map[string]interface{})["last"].(string),
					Email:          userData["email"].(string),
					Email_verified: true,
					Sub:            "sub",
				},
			},
			MediaFiles: []string{
				userData["picture"].(map[string]interface{})["large"].(string),
			},
		}
		users = append(users, user)
	}

	return users, nil

}

func generate_x_Users(amount string) ([]models.User, error) {
	return get_random_user(amount)
}

func insert_Mock_Users(amount string) (*models.User, error) {
	users, err := generate_x_Users(amount)
	if err != nil {
		return nil, err
	}

	for _, user := range users {
		//insert user into database
		_, err := DBManager.DB.Collection("users").InsertOne(context.Background(), user)
		if err != nil {
			return nil, err
		}

	}
	return &users[0], nil
}

func Gen_Mock_Users(amount string) (*models.User, error) {
	user, err := insert_Mock_Users(amount)
	if err != nil {
		fmt.Println("Error inserting mock users:", err)
		return nil, err
	}
	return user, nil
}

File: python/helpers/mock_users_gen.py
from fastapi import APIRouter, HTTPException
from typing import List
import requests
from datetime import datetime
from pydantic import BaseModel, Field
from bson import ObjectId
from helpers.db_manager import DBManager
from models.user import User, DateOfBirth, Height, MatchPreferences, Provider

# get db from db_manager
db = DBManager.db
users_collection = db["users"]



async def get_random_user(amount: str):
    url = f"https://randomuser.me/api/?results={amount}"
    response = requests.get(url)
    if response.status_code != 200:
        raise HTTPException(status_code=500, detail="Error fetching users")
    data = response.json()
    users = []
    for user_data in data["results"]:
        dob_string = user_data["dob"]["date"]
        # Adjust the format to include milliseconds
        parsed_date = datetime.strptime(dob_string, "%Y-%m-%dT%H:%M:%S.%fZ")
        dob = DateOfBirth(
            day=parsed_date.day,
            month=parsed_date.month,
            year=parsed_date.year,
        )
        id = ObjectId()
        user = User(
            id=id,
            email=user_data["email"],
            verified_email=True,
            phone_number=user_data["phone"],
            first_name=user_data["name"]["first"],
            last_name=user_data["name"]["last"],
            gender=user_data["gender"],
            location=user_data["location"]["country"],
            match_preferences=MatchPreferences(
                genders=["Men", "Women"],
                relationship_types=["Long Term Relationships"],
                minimum_age=18,
                maximum_age=39,
                maximum_distance=40,
            ),
            date_of_birth=dob,
            height=Height(feet=5, inches=8),
            providers={"google": Provider(
                name=f"{user_data['name']['first']} {user_data['name']['last']}",
                email=user_data["email"],
                email_verified=True,
                sub="sub",
            )},
            media_files=[user_data["picture"]["large"]],
        )
        users.append(user)
    return users


async def generate_x_users(amount: str):
    return await get_random_user(amount)


async def insert_mock_users(amount: str) -> User:
    users = await generate_x_users(amount)
    try :
        for user in users:
            users_collection.insert_one(user.dict())
        return users[0], None
    except Exception as e:
        return None, e



async def gen_mock_users(amount: str):
    user, err = await insert_mock_users(amount)
    return user, err


async def generate_mock_users(amount: str):
    user, err = await gen_mock_users(amount)
    return user, err
